### PR TITLE
fix: re-list when target change

### PR DESCRIPTION
**Pull request by LaurenceLiZhixin**
Force re-list pool-scope resource when proxy change between loadbalancer and pool-coordinator

### CODE REVIEW

**Comment by Congrool** [comment]
I still have some questions for such problem:
1. client sent list/watch requests and they were handled by yurthub. Then the yurthub crashed and restarted. And at this time, the handler changed(possibly because pool-coordinator become not ready), will the client relist the resource?
2. When the long time watch connection breaks causing by the server-side(e.g. cloud-edge network breaks or pool-coordinator restarted),  will yurthub always make the client relist?

**Comment by LaurenceLiZhixin** [comment]
Thanks for your great comments @Congrool 

1. After my tests, yurthub restart will cause all existing watch requests to be disconnected from client side, and the client will trigger a re-list-watch. If the restart process occurs with the flow change from edge to cloud, after the stability, yurthub got ready coordinator status, and yurthub will cut the cloud to the edge again , the client-side re-list-watch will still be triggered by the error event injection provided by this pr.

2.
- If the cloud edge network is disconnected: If yurthub is directly connected to the cloud, the edge pool-coordinator is unavailable. Therefore, there is no problem.
- If the pool-coordinator restarts, the existing watch requests are cut and the client re-list-watch. Then yurthub will determine that a pool-coordinator is abnormal, and trigger a re-list-watch for all existing watches through the error event injection provided by this pr, and then client side list-watch agent and yurthub proxy to the cloud. After the pool-coordinator restarted, the flow is changed to local and all re-list-watch are triggered again. There will be no problem after the experiment.

File: pkg/yurthub/proxy/remote/loadbalancer.go [hunk 1]
```diff
@@ -211,19 +211,21 @@ func (lb *loadBalancer) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 			for {
 				select {
 				case <-t.C:
-					if lb.coordinatorGetter == nil {
-						continue
-					}
+					klog.Infof("lb.coordinatorGetter is %s", lb.coordinatorGetter)
```
Comment (Congrool) [inline pkg/yurthub/proxy/remote/loadbalancer.go:214]: We may not need such log, please remove

File: pkg/yurthub/proxy/remote/loadbalancer.go [hunk 2]
```diff
@@ -211,19 +211,21 @@ func (lb *loadBalancer) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
...
 					coordinator := lb.coordinatorGetter()
 					if coordinator == nil {
+						klog.Infof("lb.coordinatorGetter().coordinator is nil")
```
Comment (Congrool) [inline pkg/yurthub/proxy/remote/loadbalancer.go:217]: ditto

File: pkg/yurthub/proxy/remote/loadbalancer.go [hunk 3]
```diff
@@ -211,19 +211,21 @@ func (lb *loadBalancer) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
...
 						continue
 					}
 					if _, isReady := coordinator.IsReady(); isReady {
 						klog.Infof("notified the pool coordinator is ready, cancel the req %s making it handled by pool coordinator", hubutil.ReqString(req))
+						util.ReListWatchReq(rw, req)
 						cloudServeCancel()
 						return
 					}
+					klog.Infof("coordinator.IsReady() is false")
```
Comment (Congrool) [inline pkg/yurthub/proxy/remote/loadbalancer.go:226]: ditto

**Comment by Congrool** [comment]
Sounds good to me

**Comment by Congrool** [comment]
Could you please squash your commits to make the commit message more readable?

**Comment by Congrool** [comment]
Another question, is it possible that:
list request is first handled by one handler(for example cloud APIServer), but the later watch request is handled by another handler(for pool-coordinator becoming healthy)

Will it still have some problems in such situation?

File: pkg/yurthub/proxy/remote/loadbalancer.go (hunk 1 above)
Comment (LaurenceLiZhixin) [reply pkg/yurthub/proxy/remote/loadbalancer.go:214]: Fine!

**Comment by LaurenceLiZhixin** [comment]
> Another question, is it possible that: list request is first handled by one handler(for example cloud APIServer), but the later watch request is handled by another handler(for pool-coordinator becoming healthy)
> 
> Will it still have some problems in such situation?

Yes, that's still a problem. But it's very rare. I think we can talk about it and fix it totally later.

**Comment by LaurenceLiZhixin** [comment]
I've fixed your great comments above. @Congrool

**Comment by Congrool** [comment]
LGTM as a solution for most cases, thanks for your contribution!

/lgtm

**Review by rambohe-ch — APPROVED** [review]

**Pull request merged**